datadotworld/filewriter.py

```diff
@@ -0,0 +1,129 @@
+# data.world-py
+# Copyright 2017 data.world, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the
+# License.
+#
+# You may obtain a copy of the License at
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+# implied. See the License for the specific language governing
+# permissions and limitations under the License.
+#
+# This product includes software developed at
+# data.world, Inc.(http://data.world/).
+from __future__ import absolute_import
+
+try:
+    from queue import Queue
+except ImportError:
+    from Queue import Queue
+from threading import Thread
+import json
+import requests
+from datadotworld.util import parse_dataset_key
+
+
+class DataDotWorldFileWriter:
+    def __init__(self, config, dataset_key, file_name,
+                 host="https://api.data.world/v0",
+                 timeout=None):
+        """
+        Construct a new data.world file writer - values written to the
+        `write()` method are streamed to the data.world api and written into
+        the specified file. The proper way to use this class is in a `with`
+        block:
+        >>> with DataDotWorldFileWriter(config, "user/dataset", "file") as w
+        >>>   w.write("")
+        which will ensure that at the end of the `with` block the file is
+        closed and the HTTP request completes
+        :param config: the data.world configuration object
+        :param dataset_key: key for the target dataset
+        :param file_name: name of file to write
+        :param host: hostname for the API - defaults to the standard
+        production API
+        :param timeout: how long to wait for a response when `close()` is
+        called before timing out (defaults to no timeout - the close() call
+        by default will wait indefinitely for a response)
+        """
+        self._queue = Queue(10)
+        self._sentinel = None
+        self._timeout = timeout
+        self._response_queue = Queue(1)
+
+        def put_request(body, response_queue):
+            ownerid, datasetid = parse_dataset_key(dataset_key)
+            response = requests.put(
+                "{}/uploads/{}/{}/files/{}".format(
+                    host, ownerid, datasetid, file_name),
+                data=body,
+                headers={
+                    'Authorization': 'Bearer {}'.format(config.auth_token)
+                })
+            response_queue.put(response)
+
+        body = iter(self._queue.get, self._sentinel)
+        self._thread = Thread(target=put_request,
+                              args=(body, self._response_queue))
+
+    def write(self, value):
```

Comment:
It think we could make this more idiomatic and error safe, if the writer was set in binary or character mode when it's created and `write()` did only one thing. Because:
1. Writing characters and bytes in the same session is likely to be an error
2. `json.dump(object, writer)` is easy enough and more flexible, given that `object` can be many things other than `dict`.

I agree that `dict`, `list`, etc., are common enough data structures that we may want to support more conveniently, I just would choose to keep `write()` simple and, if necessary, expose specialized write methods (e.g. `write_json()`).